File: 프로그래머스/2/340211. ［PCCP 기출문제］ 3번 ／ 충돌위험 찾기/［PCCP 기출문제］ 3번 ／ 충돌위험 찾기.js
function solution(points, routes) {
    const x = routes.length;
    const robotsPos = Array.from(Array(x), () => []);
    
    let maxRouteLen = 1;
    for(let i = 0;i < x;i++) {
        const route = routes[i];
        const len = route.length;
        
        for(let j = 1;j < len;j++) {
            const [r1, c1] = points[route[j - 1] - 1]; // from
            const [r2, c2] = points[route[j] - 1]; // to

            if(r1 > r2) {
                for(let r = r1;r > r2;r--) {
                    robotsPos[i].push([r, c1]);
                }
            } else if(r1 < r2) {
                for(let r = r1;r < r2;r++) {
                    robotsPos[i].push([r, c1]);
                }
            }
            
            if(c1 > c2) {
                for(let c = c1;c > c2;c--) {
                    robotsPos[i].push([r2, c]);
                }
            } else if(c1 < c2) {
                for(let c = c1;c < c2;c++) {
                    robotsPos[i].push([r2, c]);
                }
            }
            if(j === len - 1) {
                robotsPos[i].push([r2, c2]);
            }
        }
        
        maxRouteLen = maxRouteLen < robotsPos[i].length ? robotsPos[i].length : maxRouteLen;
    }
    
    let answer = 0;
    // 각 초마다 위치가 곂치는 것
    for(let sec = 0;sec < maxRouteLen;sec++) {
        
        // 충돌 여부 검사했는지
        let check = Array.from(Array(x), () => false);
        for(let i = 0;i < x - 1;i++) {
            if(!robotsPos[i][sec]) continue;
            if(check[i]) continue;
            
            let flag = false;
            for(let j = i + 1;j < x;j++) {
                if(!robotsPos[j][sec]) continue;
                if(check[i]) continue;
                
                const [r1, c1] = robotsPos[i][sec];
                const [r2, c2] = robotsPos[j][sec];
                
                if(r1 === r2 && c1 === c2) {
                    check[j] = true;
                    flag = true;
                }
            }
            if(flag) answer++;
        }
    }
    
    return answer;
}

File: 프로그래머스/2/340211. ［PCCP 기출문제］ 3번 ／ 충돌위험 찾기/［PCCP 기출문제］ 3번 ／ 충돌위험 찾기.java
import java.util.*;

class Solution {
    
    public int solution(int[][] points, int[][] routes) {
        int n = points.length;
        int x = routes.length;
        
        Map<String, Integer> routeMap = new HashMap<>();
        
        for(int[] route: routes) {
            int len = route.length;
            
            int time = 0;
            for(int j = 0;j < len - 1;j++) {
                int[] cur = points[route[j] - 1];
                int[] next = points[route[j + 1] - 1];
                
                if(cur[0] > next[0]) {
                    for(int r = cur[0];r > next[0];r--) {
                        String key = makeKey(r, cur[1], time);
                        boolean has = routeMap.containsKey(key);
                        routeMap.put(key, has ? routeMap.get(key) + 1 : 1);
                        time++;
                    }
                } else if(cur[0] < next[0]) {
                    for(int r = cur[0];r < next[0];r++) {
                        String key = makeKey(r, cur[1], time);
                        boolean has = routeMap.containsKey(key);
                        routeMap.put(key, has ? routeMap.get(key) + 1 : 1);
                        time++;
                    }
                }
                
                if(cur[1] > next[1]) {
                    for(int c = cur[1];c > next[1];c--) {
                        String key = makeKey(next[0], c, time);
                        boolean has = routeMap.containsKey(key);
                        routeMap.put(key, has ? routeMap.get(key) + 1 : 1);
                        time++;
                    }
                } else if(cur[1] < next[1]) {
                    for(int c = cur[1];c < next[1];c++) {
                        String key = makeKey(next[0], c, time);
                        boolean has = routeMap.containsKey(key);
                        routeMap.put(key, has ? routeMap.get(key) + 1 : 1);
                        time++;
                    }
                }
                if(j == len - 2) {
                    String key = makeKey(next[0], next[1], time);
                    boolean has = routeMap.containsKey(key);
                    routeMap.put(key, has ? routeMap.get(key) + 1 : 1);
                }
            }
        }
        
        int answer = 0;
        for(Integer count: routeMap.values()) {
            if(count > 1) answer++;
        }
        return answer;
    }
    
    public String makeKey(int r, int c, int time) {
        return r + "," + c + "," + time;
    }
}